README.md
# Funky Main App

Funky application responsible for start DaFunk ecosystem.

## What's inside?
- A small Ruby application that uses [da_funk](https://github.com/cloudwalkio/da_funk).
- Example test cases.
- All the scripts needed to make it work.

## Flags from params.dat

- `emv_application_name_table` - If 1 display application name by params.dat;
- `connection_management` (Default 1) - If 1 connection management will be turned on and the main application will keep the communication up based on the last configuration. If any communication issue the main application will attempt in loop;
- `conn_fallback_drops_limit` (Default 0: disable) - Drop limit of primary connection to start a fallback;
- `conn_fallback_config` (Default "": disable) - Fallback configuration, ei.: `GPRS|chili|chili|`;
- `conn_fallback_timer` (Default 0: disable) - Time until try the primary connection again (seconds);
- `api_token` (Default "": disable) - Manager api token to upload any log file;
- `disable_datetime` (Default 0: disable) - If 1 Disable date time display at main idle screen;
- `emv_enabled` (Default 0: disable) - If 1 module EMV will be loaded and EMV transactions will be accepted;
- `transaction_conn_check` (Default 0: disable) - If 1 check if terminal is connected to perform transactions (Mag or EMV) and reversals;
- `access_token` (Default "": disable) - Payment channel access token;
- `payment_channel_enabled` (Default 0: disable) - If 1 payment channel will be handle by the main application, access_token must to exist.
- `encrypt_card` (Default 0: disable) - If 1 the PAN will be encrypted using 3DES injected keys to share between modules;
- `emv_application` (Default "": disable) - Application that will be called after a EMV(insert card) input;
- `attach_gprs_timeout` (Default "": disable) - timeout in miliseconds
- `backlight_control` (Default "120": enable, seconds) - If enable ( > 0) perform backlight control and turn off the backlight based on the time in seconds of this parameter if any event happen, if disable (0) keep backlight always on;
- `countdown_application` (Default "": AdministrationMenu) - Application to be called if communication failure countdown;
- `countdown_max_timeout` (Default "5") - countdown timeout;
- `update_interval` (Default "24") - Applications update check interval;
- `system_update_interval` (Default "168") - Runtime update check interval;
- `payment_channel_check_limit` (Default "0": disable) - Enable(1) / Disable(0) daily payment channel limit;
- `payment_channel_limit` (Default "0") - Configure daily limit to payment channel creation;
- `transaction_http_enabled` (Default "1") - Enabled http transaction interface, if "0" PaymentChannel via Websocket;
- `log_upload_enabled` (Default "0") - Upload communication and key logs;
- `emv_application_name_image` (Default "0") - If "1" enable image display for emv application selection, it's required 3 images on ./shared path: `emv_selection_credit_debit.bmp`, `emv_selected_credit.bmp` and `emv_selected_debit.bmp`.

## Setup

1. Install Ruby
2. Bundle `bundle install`

## Running

Make sure you have either mruby or the CloudWalk CLI tool, then run `rake` to build the app, and one of the following commands to see it working, depending on your interests:

- `cloudwalk emulate`: to run the application in a graphical emulator in a separated window.
- `cloudwalk run`: to run the application in text mode in the same window.

## Deploying

In case you're using this skeleton from a new application created by our CLI, your application will be deployed to our servers after `git push`. Be sure to add all the files you require and to test everything before pushing, otherwise the deploy will be rejected.

## Test

1. Unit `rake test:unit`
2. Integration `rake test:integration`
3. All `rake test`

## Customizing the app

To customize the application, read our docs at <https://docs.cloudwalk.io/en/cli> or check [da_funk's source code](https://github.com/cloudwalkio/da_funk).

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Added some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new Pull Request

## License

This project is released under the [MIT License](https://opensource.org/licenses/MIT).


RELEASE_NOTES.md
# Main

Funky application responsible for start DaFunk ecosystem.

### 3.66.0 - 2021-04-27

- Update cloudwalk_handshake (1.22.1);
- Added new handler emv_table_update to deal with emv parameters check.

### 3.65.2 - 2021-03-11

- Rescue exceptions on ruby apps tasks schedule.

### 3.65.1 - 2021-02-26

- Added new approach to send large log files divided in parts.

### 3.65.0 - 2021-02-19

- Added new handler to check if emv_table_reload file exists, it means InfinitePay payment application has updated the table and main application needs to reload it;
- Map menu# button to access payment application, only for versions >= 8.1 and S920 model;
- Added :main_menu option which access main menu without password, because in this case the password was validated in another process;
- Added possibility to schedule tasks in minutes and hours for ruby applications;
- Update da_funk (3.33.1)

### 3.64.0 - 2021-02-08

- Update da_funk (3.33.0);
- Added support to images on CloudwalkUpdate#system and CloudwalkUpdate#application;
- Added support to images on system update process.

### 3.63.3 - 2021-01-20

- Added PATH_UPDATE_DONE constant on SystemUpdate class;
- Delete PATH_UPDATE_DONE file if device is up to date.

### 3.63.2 - 2021-01-14

- Added log of pusbsub id registered;
- Close socket on communication thread after media switch only in case media changed and socket is connected.

### 3.63.1 - 2021-01-07

- Added workaround to solve system update error. There's a bug in the PAX::System#update method which returns a wrong result that causes an error in the system update flow, this change workaround this error.

### 3.63.0 - 2021-01-04

- Update keyboard screens;
- Call to thread pause before configuring communication;
- Reload metadata after switching communication;
- Added log in order to know that communication config was updated;
- Set statusbar attributes as nil if thread is paused;
- Update da_funk (3.32.0).

### 3.62.1 - 2020-11-27

- Send all logs when executed manually.

### 3.62.0 - 2020-11-27

- Added new battery percentage icons;
- Update cloudwalk_handshake (1.22.0);
- Enable touchscreen application config on 8.X.X runtime version only;
- Update da_funk (3.31.0).

### 3.61.0 - 2020-11-16

- Added new icon for 3G connection;
- Added new s920 main screen image;
- Added support to touchscreen events which executes other apps;
- Added support to send additional parameters in the function executed from scheduler;
- Update da_funk (3.30.0).

### 3.60.2 - 2020-11-04

- Update cloudwalk_handshake (1.21.3);
- Update da_funk (3.29.1).

### 3.60.1 - 2020-10-26

- Added device was restarted info on log.

### 3.60.0 - 2020-10-07

- Update da_funk (3.28.2);
- Update some images;
- Added new handler to check params.dat status each 60 minutes;
- Update cloudwalk_handshake (1.21.1).

### 3.59.3 - 2020-09-29

- Concatenate and send all logs at once;
- Fixed filename. Log txt should have the date of the day that is being sent;
- Do not send full path as LogsMenu#send_file is already adding path;
- Send all log files separated;
- Moved logs ui feedback to LogsMenu#send_file_menu and LogControl#upload;
- LogsMenu#send_file return booleam as result;
- Store return of each log send.

### 3.59.2 - 2020-09-29

- Update cloudwalk_handshake (1.21.0)

### 3.59.1 - 2020-09-26

- Do not try to download system parts if update file does not exists;
- Check if update dat object exists on SystemUpdate#done? before start it;
- Update da_funk (3.28.1).

### 3.59.0 - 2020-09-25

- Add main application version on config.dat file;
- Move thread loop process to Main#thread_loop metohod, also added rescue on Main#thread_loop;
- Update cloudwalk_handshake (1.20.1);
- Update da_funk (3.27.0);
- Improved system update messages;
- Show system update message at line 0 and alway clear it before;
- Added system_update_start message;
- Check system update status at zero position of string array;
- Print system update message after device restart, in case system update is in progress;
- Added additional information in the system update file in order to know that the device needs to restart before proceed with the system update;
- Added CloudwalkUpdate#wait_connection method;
- Added CloudwalkUpdate#count_down method;
- Added CloudwalkUpdate#system_in_progress? method;
- Wait terminal to connect before start system update in case device has restarted;
- Added system update management file on CloudwalkUpdate;
- Fixed display messages on application update proccess;
- Update da_funk (3.28.0).

### 3.58.0 - 2020-09-14

- Update cloudwalk_handshake (1.20.0);
- Update da_funk (3.26.0);
- Update virtual keyboard images.

### 3.57.0 - 2020-08-28

- Update da_funk (3.25.1);

lib/cloudwalk_setup.rb
class CloudwalkSetup
  include DaFunk::Helper

  def self.boot(start_attach = true)
    boot_layout
    self.setup_notifications
    self.setup_listeners
    self.setup_events
    LogControl.purge
    CloudwalkFont.setup
    PosxmlParser.setup
    BacklightControl.setup
    DaFunk::ParamsDat.parameters_load
    self.schedule_routines_from_rb_apps
    self.setup_keyboard_events_from_rb_apps
    self.setup_touchscreen_events_from_rb_apps
    self.pre_load_applications
    DaFunk::EventHandler.new :magnetic, nil do end
    if update_process_in_progess?
      MediaConfiguration.gprs_default({
        apn: 'cloudwalk.algar.br',
        user: 'algar',
        pass: '1212'
      })
    end
    Context::ThreadScheduler.start
    Device::Setting.update_attributes('main_app_version' => Main.version)
    if update_process_in_progess? || !AdminConfiguration.device_activated?
      AdminConfiguration.configure_payment_application
    end
  end

  def self.setup_listeners
    DaFunk::EventListener.new :key_main do |event|
      event.check do
        key = getc(100)
        if key != Device::IO::KEY_TIMEOUT
          BacklightControl.on
          handler = event.handlers[key]
          if handler
            handler.perform
            BacklightControl.on
          end
        end
      end
    end

    DaFunk::EventListener.new :touchscreen do |event|
      event.check do
        touch_clear
        x, y = getxy_stream(800)
        if x && y
          event.handlers.each do |option, handler|
            if option.is_a?(Hash)
              if option.include?(:x) && option.include?(:y)
                if option[:x].include?(x) && option[:y].include?(y)
                  BacklightControl.on
                  Device::Audio.beep(7, 60)
                  handler.perform
                end
              end
            end
          end
        end
      end
    end

    DaFunk::EventListener.new :magnetic do |event|
      event.start do
        @mag = Device::Magnetic.new
        true
      end

      event.check do
        @mag = Device::Magnetic.new unless @mag.open?
        if @mag.open? && @mag && @mag.swiped?
          DaFunk::PaymentChannel.connect(false) if DaFunk::PaymentChannel.channel_limit_exceed?
          handler = event.handlers.find { |option, h| @mag.bin?(h.option) }
          if handler
            BacklightControl.on
            if check_connection
              handler[1].perform(@mag)
            end
            BacklightControl.on
          end
          DaFunk::PaymentChannel.close! if DaFunk::PaymentChannel.transaction_http?
          event.finish
          event.start
        end
      end

      event.finish do
        @mag.close
        EmvTransaction.reboot
      end
    end

    DaFunk::EventListener.new :schedule do |event|
      event.check do
        handler = event.handlers.find { |option, h| h.execute? }
        handler[1].perform if handler
      end
    end

    DaFunk::EventListener.new :emv do |event|
      event.start do
        EmvTransaction.boot
        true
      end

      event.check do
        if EmvTransaction.opened? && DaFunk::ParamsDat.file["emv_enabled"] != "0"
          if EmvTransaction.detected?
            if DaFunk::PaymentChannel.channel_limit_exceed?
              DaFunk::PaymentChannel.connect(false)
            end
            BacklightControl.on
            if CloudwalkSetup.check_connection
              EmvTransaction.clean
              EmvTransaction.initialize do |emv|
                FunkyEmv::Ui.display(:emv_processing, :line => 2, :column => 1)
                emv.set_initial_data
                handler = event.handlers.first
                handler[1].perform(emv.select) if handler && handler[1]
              end
              EmvTransaction.reboot
            end
            DaFunk::PaymentChannel.close! if DaFunk::PaymentChannel.transaction_http?
            BacklightControl.on
          end
        else
          EmvTransaction.boot if DaFunk::ParamsDat.file["emv_enabled"] != "0"
        end
      end

      event.finish do
        EmvTransaction.clean
      end
    end

    DaFunk::EventListener.new :file_exists do |event|
      event.start do
        true
      end

      event.check do
        event.handlers.each do |option, handler|
          handler.perform if File.exists?(option)
        end
      end
    end

    DaFunk::EventListener.new :boot do |event|
      event.start do true end

      event.check do
        event.handlers.each do |file_path, handler|
          handler.perform if File.exists?(file_path)
        end
        Device::Setting.boot = "0"
        event.delete
      end
    end

    DaFunk::EventHandler.new :file_exists, "./shared/system_update" do
      content = File.read('shared/system_update').split("\n")
      CloudwalkUpdate.system if content[0] == 'DONE'
    end

    DaFunk::EventHandler.new :file_exists, "./shared/application_update" do
      CloudwalkUpdate.application
    end

    # Because payment application can update the table and it's running in other runtime instance
    DaFunk::EventHandler.new :file_exists, './shared/emv_table_reload' do
      if EmvTransaction.opened?
        EmvTransaction.load('4')
      else
        EmvTransaction.boot
      end
      File.delete('./shared/emv_table_reload') if File.exists?('./shared/emv_table_reload')
    end

    DaFunk::EventListener.new :file_exists_once do |event|
      event.start do @file_exists_once = {}; true end

      event.check do
        event.handlers.each do |option, handler|
          if @file_exists_once[option].nil? && File.exists?(option)
            @file_exists_once[option] = true
            handler.perform
          end
        end
      end
    end

    DaFunk::EventListener.new :payment_channel do |event|
      event.start do
        @id ||= Context::ThreadPubSub.subscribe
        true
      end

      event.check do
        if Context::ThreadPubSub.listen(@id) == "communication_update"
          Device::Runtime.system_reload
        end

        if (! ThreadScheduler.pause?(ThreadScheduler::THREAD_EXTERNAL_COMMUNICATION, 200) &&
            payload = DaFunk::PaymentChannel.current.check)
          payload, notification = DaFunk::Notification.check(payload)
          if notification && notification.reply
            DaFunk::PaymentChannel.current.write(notification.reply)
          end
          handler = event.handlers[payload]
          handler.perform(notification) if handler
        end
      end
    end

    DaFunk::EventHandler.new :schedule, hours: 2, slot: 'battery_check' do
      if Device::System.battery < 20 && ! Device::System.power_supply
        if File.exists?('./shared/battery_low.bmp')
          Device::Display.print_bitmap('./shared/battery_low.bmp')
        else
          Device::Display.clear
          Device::Display.print_line(I18n.t(:battery_low).split("\n")[0], 3, 0)
          Device::Display.print_line(I18n.t(:battery_low).split("\n")[1], 4, 0)
        end

        5.times do
          Device::Audio.beep(0, 180)
          getc(1500)
        end
      end
    end

    DaFunk::EventHandler.new :schedule, minutes: 60 do
      if DaFunk::ParamsDat.corrupted?
        ContextLog.info "params.dat file was corrupted"
        ContextLog.info "Starting download of new params.dat"
        DaFunk::ParamsDat.restart if DaFunk::ParamsDat.download
      end
    end

    DaFunk::EventListener.new :emv_table_update do |event|
      event.start do
        true
      end

      event.check do
        handler = event.handlers.find { |option, h| h.execute? }
        handler[1].perform if handler
      end
    end

    if DaFunk::ParamsDat.file["emv_table_update_interval"].to_s.empty?
      interval = 168
    else
      interval = DaFunk::ParamsDat.file["emv_table_update_interval"].to_i
    end
    app = DaFunk::ParamsDat.file["emv_table_update_app_name"]
    DaFunk::EventHandler.new :emv_table_update, hours: interval, slot: 'emv_table_update_interval' do
      json = { initialize: 'emv_table' }.to_json
      if !app.to_s.empty?
        Device::Runtime.execute(app, json)
      else
        application = DaFunk::ParamsDat.ruby_executable_apps.find do |app|
          app.name == 'cw_infinitepay_app'
        end
        application.execute(json) if application
      end
    end
  end

  def self.setup_communication_listeners
    DaFunk::EventListener.new :payment_channel do |event|
      event.start do
        DaFunk::PaymentChannel.check(false)
        true
      end

      event.check do
        if (! ThreadScheduler.pause?(ThreadScheduler::THREAD_EXTERNAL_COMMUNICATION, 200) &&
            payload = DaFunk::PaymentChannel.check(false))
          handler = event.handlers[payload]
          if handler
            handler.perform
          else
            Context::ThreadChannel.write(:recv, payload) if payload.is_a?(String)
          end
        end
      end
    end

    # Necessary to enable Listener check
    DaFunk::EventHandler.new :payment_channel, :nothing do
      true
    end

    DaFunk::EventListener.new :communication do |event|
      event.start do
        true
      end

      event.check do
        handler = event.handlers[DaFunk::ConnectionManagement.check]
        handler.perform if handler
      end
    end

    DaFunk::EventHandler.new :communication, :attach_registration_fail do
      attach(print_last: false)
    end

    DaFunk::EventHandler.new :communication, :fallback_communication do
      if DaFunk::ConnectionManagement.fallback_valid?
        DaFunk::PaymentChannel.close!
        Device::Network.shutdown
        if DaFunk::ConnectionManagement.recover_fallback
          attach(print_last: false)
        end
      end
    end

    DaFunk::EventHandler.new :communication, :primary_communication do
      if DaFunk::ConnectionManagement.fallback_valid?
        DaFunk::PaymentChannel.close!
        Device::Network.shutdown
        if DaFunk::ConnectionManagement.recover_primary
          unless attach(print_last: false)
            Device::Network.shutdown
            if DaFunk::ConnectionManagement.recover_fallback
              attach(print_last: false)
            end
          end
        end
      end
    end

    DaFunk::EventListener.new :schedule do |event|
      event.start do
        true
      end

      event.check do
        handler = event.handlers.find { |option, h| h.execute? }
        handler[1].perform if handler
      end
    end

    value = DaFunk::ParamsDat.file["system_update_interval"]
    if value.to_s != "0"
      interval = (value.to_s.empty? ? 360 : value.to_i)
      DaFunk::EventHandler.new :schedule, hours: interval, slot: "system_update_interval" do
        if DaFunk::ParamsDat.file["system_update_thread_enable"] == "1"
          SystemUpdate.bg_start unless SystemUpdate.current&.done?
        end
      end
    end

    DaFunk::EventListener.new :background_system_update do |event|
      event.start do true end

      event.check do
        if SystemUpdate.current
          handler = event.handlers.first
          handler[1].perform if handler
        end
      end
    end

    DaFunk::EventHandler.new :background_system_update, :nothing do
      SystemUpdate.current&.bg_check
    end

    DaFunk::EventListener.new :file_exists do |event|
      event.start do
        true
      end

      event.check do
        event.handlers.each do |option, handler|
          handler.perform if File.exists?(option)
        end
      end
    end

    DaFunk::EventHandler.new :file_exists, "./shared/system_update" do
      SystemUpdate.bg_start
    end

    DaFunk::EventHandler.new :schedule, minutes: 2 do
      GC.start
    end
  end

  def self.countdown_menu
    if timeout = DaFunk::ParamsDat.file["countdown_max_timeout"] && timeout.integer?
      timeout = timeout.to_i
    end
    (1..(timeout || 5)).to_a.reverse.each do |second|
      DaFunk::PaymentChannel.print_info(I18n.t(:attach_registration_fail, :args => second), true)
      key = getc(1000)
      if key == Device::IO::ENTER
        if (app = DaFunk::ParamsDat.file["countdown_application"])
          Device::Runtime.execute(app)
        else
          AdminConfiguration.perform
        end
      elsif key == Device::IO::F1 || key == Device::IO::FUNC
        AdminConfiguration.perform
      end
    end
  end

  def self.setup_keyboard_events_from_rb_apps
    DaFunk::ParamsDat.ruby_executable_apps.each do |app|
      if File.exists?("#{app.dir}/cw_keys.json")
        app_keys = JSON.parse(File.read("#{app.dir}/cw_keys.json"))
        app_keys.each do |key, options|
          DaFunk::EventHandler.new :key_main, key do
            app       = options["app"]
            operation = options["initialization"]
            Device::Runtime.execute(app, operation.to_json)
          end
        end
      end
    end
  end

  def self.setup_touchscreen_events_from_rb_apps
    # This check of version it's because versions < 8 cannot run reversal through notification in the home screen
    major, min, patch = Device.version.to_s.split('.').map { |v| v.to_i }
    if major >= 8
      DaFunk::ParamsDat.ruby_executable_apps.each do |app|
        if File.exists?("#{app.dir}/cw_touchscreen.json")
          functions = JSON.parse(File.read("#{app.dir}/cw_touchscreen.json"))
          functions.each do |function, options|
            range = options['range'].to_a.inject({}) do |result, value|
              ary = value[1].split('..').map { |v| v.to_i }
              result[:x] = ary[0]..ary[1] if value[0] == 'x'
              result[:y] = ary[0]..ary[1]
              result
            end
            DaFunk::EventHandler.new :touchscreen, range do
              Device::Runtime.execute(options['app'], {initialize: function}.to_json)
            end
          end
        end
      end
    end
  end

  def self.schedule_routines_from_rb_apps
    DaFunk::ParamsDat.ruby_executable_apps.each do |app|
      schedule_path = "#{app.dir}/cw_app_schedule.json"
      if File.exists?(schedule_path)
        schedule_params = JSON.parse(File.read(schedule_path))
        schedule_params["routines"].each do |params|
          schedule_interval = {}
          ruby_app      = params["app"]
          function      = {
                            initialize: params["routine"]["initialize"],
                            parameters: params["routine"]["parameters"]
                          }
          if params["routine"]["type_time"] == 'hours' && params["routine"]["slot"]
            schedule_interval[:slot]  = params["routine"]["slot"]
            schedule_interval[:hours] = params["routine"]["interval"].to_i
          elsif params["routine"]["type_time"] == 'minutes'
            schedule_interval[:minutes] = params["routine"]["interval"].to_i
          else
            schedule_interval[:seconds] = params["routine"]["interval"].to_i
          end
          file_check    = params["routine"]["file_check"]
          function_boot = { :initialize => params["routine"]["boot"] }

          DaFunk::EventHandler.new :schedule, schedule_interval do
            Device::Runtime.execute(ruby_app, function.to_json)
          end

          if file_check
            DaFunk::EventHandler.new :file_exists, "./shared/#{file_check}" do
              Device::Runtime.execute(ruby_app, function.to_json)
            end
          end

          if params["routine"].include?("boot")
            DaFunk::EventHandler.new :boot, ruby_app do
              Device::Runtime.execute(ruby_app, function_boot.to_json)
            end
          end
        end
      end
    end
  rescue => e
    ContextLog.exception(e, e.backtrace)
  end

  def self.setup_events
    major, min, patch = Device.version.to_s.split('.').map { |v| v.to_i }

    if InputTransactionAmount.enabled? && InputTransactionAmount.emv_ctls_table_installed?
      DaFunk::EventHandler.new :touchscreen, {:x => 41..199, :y => 179..233} do
        InputTransactionAmount.call
      end

      (1..9).to_a.each do |key|
        DaFunk::EventHandler.new :key_main, key.to_s do InputTransactionAmount.call(key.to_s) end
      end
    end

    DaFunk::EventHandler.new :key_main, Device::IO::ENTER do CloudwalkSetup.start            end
    DaFunk::EventHandler.new :key_main, Device::IO::F1    do AdminConfiguration.perform      end
    if Device::System.model == "link2500"
      DaFunk::EventHandler.new :key_main, Device::IO::ALPHA    do AdminConfiguration.perform end
    end

    if Device::System.model == "s920"
      if (major == 8 && min >= 1) || major > 8
        DaFunk::EventHandler.new :key_main, Device::IO::ALPHA  do CloudwalkSetup.start      end
      else
        DaFunk::EventHandler.new :key_main, Device::IO::ALPHA  do AdminConfiguration.perform      end
      end
    end

    DaFunk::EventHandler.new :key_main, Device::IO::CLEAR do Device::Printer.paperfeed       end
    if Context.development?
      DaFunk::EventHandler.new :key_main, Device::IO::F2    do DaFunk::Engine.stop!          end
      if Device::System.model != "link2500"
        if Device::System.model == "s920"
          DaFunk::EventHandler.new :key_main, Device::IO::FUNC do DaFunk::Engine.stop!        end
        else
          DaFunk::EventHandler.new :key_main, Device::IO::ALPHA do DaFunk::Engine.stop!        end
        end
      end
    end

    DaFunk::EventHandler.new :payment_channel, :notification do |notification|
      BacklightControl.on
      notification.perform
      BacklightControl.on
    end

    value = DaFunk::ParamsDat.file["update_interval"]
    interval = (value.to_s.empty? ? 120 : value.to_i)
    DaFunk::EventHandler.new :schedule, hours: interval, slot: "update_interval" do
      CloudwalkUpdate.application
    end

    value = DaFunk::ParamsDat.file["system_update_interval"]
    if value.to_s != "0"
      interval = (value.to_s.empty? ? 360 : value.to_i)
      DaFunk::EventHandler.new :schedule, hours: interval, slot: "system_update_interval" do
        if DaFunk::ParamsDat.file["system_update_main_enabled"] == "1"
          CloudwalkUpdate.system
        end
      end
    end

    value = DaFunk::ParamsDat.file["log_send_interval"] || 24
    DaFunk::EventHandler.new :schedule, hours: value, slot: "log" do
      BacklightControl.on
      LogControl.upload
    end

    DaFunk::EventHandler.new :schedule, minutes: 2 do
      GC.start
    end
  end

  def self.setup_notifications
    DaFunk::NotificationCallback.new "APP_UPDATE", :on => Proc.new { DaFunk::ParamsDat.update_apps(true); Device::System.reboot }
    DaFunk::NotificationCallback.new "SETUP_DEVICE_CONFIG", :on => Proc.new { DaFunk::ParamsDat.update_apps(true) }
    DaFunk::NotificationCallback.new "RESET_DEVICE_CONFIG", :on => Proc.new { DaFunk::ParamsDat.format! }
    DaFunk::NotificationCallback.new "REBOOT", :on => Proc.new { Device::System.reboot }

    DaFunk::NotificationCallback.new "SYSTEM_UPDATE", :on => Proc.new { SystemUpdate.new.start }
    DaFunk::NotificationCallback.new "CANCEL_SYSTEM_UPDATE", :on => Proc.new { }
    DaFunk::NotificationCallback.new "TIMEZONE_UPDATE", :on => Proc.new { Device::Setting.cw_pos_timezone = "" }
    DaFunk::NotificationCallback.new "SHOW_MESSAGE", :on => Proc.new { |message, datetime|
      Device::Display.clear
      date = datetime.sub(" ", "-").split("-")
      Device::Display.print_line("#{date[1]}/#{date[0]}/#{date[2]} #{date[3]}", 0)
      Device::Display.print_line("#{message}", 2)
      getc(0)
    }
    DaFunk::NotificationCallback.new "PROCESSING", :on => Proc.new { |app,params|
      file, ext = app.split(".")
      if ext == "posxml"
        FileDb.new("./shared/#{file}.dat")["notification"] = params
        Device::Runtime.execute(app)
      else
        Device::Runtime.execute(app, params)
      end
    }
  end

  def self.check_connection
    if DaFunk::ParamsDat.file["transaction_conn_check"] == "1"
      if Device::Network.connected?
        true
      else
        Device::Display.clear
        I18n.pt(:transaction_no_connection)
        getc(5000)
        false
      end
    else
      true
    end
  end

  def self.execute(application = nil)
    application ||= DaFunk::ParamsDat.executable_app
    unless application
      application = DaFunk::ParamsDat.application_menu
    end
    application.execute if application
  end

  def self.start
    applications = DaFunk::ParamsDat.executable_apps
    application = DaFunk::ParamsDat.executable_app
    if DaFunk::ParamsDat.exists?
      if (applications && applications.size > 1) || (application && application.exists?)
        return self.execute
      end
    end
    CloudwalkWizard.new.start
  end

  def self.pre_load_applications
    DaFunk::ParamsDat.ruby_executable_apps.each do |application|
      application.start
    end
  end

  def self.boot_layout
    boot_layout_file = {
      :default =>         './shared/boot_welcome.bmp',
      :update_process => './shared/six_steps_updating.bmp',
      :attach_network => './shared/network_conectar_init1.bmp'
    }

    I18n.configure("main", Device::Setting.locale)

    if update_process_in_progess?
      Device::Display.print_bitmap(boot_layout_file[:update_process])
    elsif CloudwalkUpdate.system_in_progress?
      if File.exists?(boot_layout_file[:attach_network])
        Device::Display.print_bitmap(boot_layout_file[:attach_network])
      else
        Device::Display.clear
        I18n.pt(:system_update, :line => 0)
      end
    else
      Device::Display.print_bitmap(boot_layout_file[:default])
    end

    Device::Setting.boot = "1"
    ContextLog.info "[I] Device was restarted"
  end

  def self.update_process_in_progess?
    File.exists?('./shared/cw_update_steps.dat')
  end
end
